File: framework/python/betterportal/bootstrap.py
"""Authenticated bootstrap state and persistent signing identity, independent of hosting."""
from __future__ import annotations

import asyncio
import base64
import json
import secrets
from typing import Any

import anyvali as av
from cryptography.exceptions import InvalidTag
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.scrypt import Scrypt

from .contracts import contract, parse
from .encryption import _decode, _encode
from .jsoncodec import loads
from .keys import secure_endpoint
from .security import KeyPair
from .storage import StateStore

_LIMIT = 1024 * 1024


class BootstrapError(ValueError):
    """Invalid protected state; messages do not disclose credentials or key material."""


def _identity(value: dict[str, Any]) -> KeyPair:
    pair = KeyPair(value["privateKeyPem"], value["kid"])
    # Node persists SPKI. Comparing the whole normalized PEM also rejects appended
    # private material which permissive PEM parsers can otherwise ignore.
    if "".join(value["publicKeyPem"].split()) != "".join(pair.public_key_pem.split()):
        raise BootstrapError("Signing identity public/private keys do not match")
    return pair


def _state(value: Any) -> dict[str, Any]:
    try:
        result = parse("BootstrapStateSchema", value)
        for field in ("cpUrl", "cpJwksUri"):
            if field in result: secure_endpoint(result[field], allow_query=field == "cpJwksUri")
        if "apiKey" in result and (not result["apiKey"].isascii() or any(ord(c) < 33 or ord(c) == 127 for c in result["apiKey"])):
            raise ValueError()
        if "identity" in result: _identity(result["identity"])
        return result
    except (ValueError, TypeError, av.ValidationError):
        raise BootstrapError("Invalid bootstrap state") from None


class BootstrapCipher:
    """Node-compatible whole-state AES-GCM. Keep the master key in a protected host secret."""
    def __init__(self, key: str):
        try:
            if not isinstance(key, str) or not key.startswith("bp_bsk_") or len(_decode(key[7:], url=True)) != 32:
                raise ValueError()
            self._cipher = AESGCM(Scrypt(salt=b"bp-bootstrap-state-v1", length=32, n=16384, r=8, p=1).derive(key.encode()))
        except ValueError:
            raise BootstrapError("Invalid bootstrap master key") from None

    @staticmethod
    def generate_key() -> str: return "bp_bsk_" + _encode(secrets.token_bytes(32))

    def encrypt(self, value: Any) -> bytes:
        value = _state(value)
        plain = json.dumps(value, ensure_ascii=False, separators=(",", ":"), allow_nan=False).encode()
        if len(plain) > _LIMIT: raise BootstrapError("Bootstrap state exceeds 1 MiB")
        iv = secrets.token_bytes(12)
        encrypted = self._cipher.encrypt(iv, plain, None)
        envelope = {"v": 1, **{k: base64.b64encode(v).decode() for k, v in {"iv": iv, "tag": encrypted[-16:], "ct": encrypted[:-16]}.items()}}
        return json.dumps(parse("BootstrapStateEnvelopeSchema", envelope), separators=(",", ":")).encode()

    def decrypt(self, value: bytes) -> dict[str, Any]:
        try:
            if len(value) > 2 * _LIMIT: raise ValueError()
            envelope = parse("BootstrapStateEnvelopeSchema", loads(value.decode("utf-8")))
            iv, tag, ct = (_decode(envelope[k]) for k in ("iv", "tag", "ct"))
            if len(iv) != 12 or len(tag) != 16 or len(ct) > _LIMIT: raise ValueError()
            return _state(loads(self._cipher.decrypt(iv, ct + tag, None).decode("utf-8")))
        except (ValueError, TypeError, InvalidTag, av.ValidationError):
            raise BootstrapError("Bootstrap state authentication failed") from None


class BootstrapStateStore:
    """One owner per state store. Each operation reloads state; failed writes never publish it.

    Cancellation belongs to the calling task. No background work or plaintext cache is retained.
    """
    def __init__(self, store: StateStore, key: str):
        self._store, self._cipher = store, BootstrapCipher(key)
        self._gate = asyncio.Lock()

    async def _read(self) -> dict[str, Any]:
        raw = await self._store.load()
        return {"version": 1} if raw is None else self._cipher.decrypt(raw)

    async def read(self, *, redacted: bool = False) -> dict[str, Any]:
        async with self._gate:
            value = await self._read()
            if not redacted: return value
            paths = []
            def redact(path, _value):
                paths.append(path)
                return "encrypted:redacted"
            # AnyVali requires encrypted markers during traversal. Replace only the
            # paths it identified in this canonical object (which has no unions).
            result = av.encrypt(contract("BootstrapStateSchema"), value, redact)
            for path in paths:
                parent = result
                for part in path[:-1]: parent = parent[part]
                parent[path[-1]] = "__redacted__"
            return result

    async def write(self, patch: dict[str, Any]) -> dict[str, Any]:
        patch = _state({"version": 1, **patch})
        async with self._gate:
            value = _state({**await self._read(), **patch})
            await self._store.save(self._cipher.encrypt(value))
            return value

    async def clear(self) -> None:
        async with self._gate:
            await self._store.save(self._cipher.encrypt({"version": 1}))

    async def identity(self) -> KeyPair:
        async with self._gate:
            value = await self._read()
            if "identity" in value: return _identity(value["identity"])
            pair = KeyPair.generate()
            value["identity"] = {"privateKeyPem": pair.private_key_pem, "publicKeyPem": pair.public_key_pem, "kid": pair.kid}
            await self._store.save(self._cipher.encrypt(value))
            return pair

File: framework/dotnet/BetterPortal/Bootstrap.cs
using AnyVali;
using Microsoft.IdentityModel.Tokens;
using Org.BouncyCastle.Crypto;
using Org.BouncyCastle.Crypto.Generators;
using System.Security.Cryptography;
using Node = System.Collections.Generic.Dictionary<string, object?>;

namespace BetterPortal;

public sealed class BootstrapException(string message) : Exception(message);

/// <summary>Node-compatible whole-state AES-GCM. Supply the master key from a protected host secret.</summary>
public sealed class BootstrapCipher
{
    private readonly byte[] key;
    public BootstrapCipher(string masterKey)
    {
        try
        {
            if (!masterKey.StartsWith("bp_bsk_", StringComparison.Ordinal) || Envelope.Decode(masterKey[7..], url: true).Length != 32)
                throw new ArgumentException();
            key = SCrypt.Generate(Envelope.Utf8.GetBytes(masterKey), "bp-bootstrap-state-v1"u8.ToArray(), 16384, 8, 1, 32);
        }
        catch (Exception error) when (error is ArgumentException or ConfigEncryptionException)
        { throw new BootstrapException("Invalid bootstrap master key"); }
    }
    public static string GenerateKey() => "bp_bsk_" + Base64UrlEncoder.Encode(RandomNumberGenerator.GetBytes(32));

    internal static KeyPair Identity(Node value)
    {
        var pair = new KeyPair((string)value["privateKeyPem"]!, (string)value["kid"]!);
        // Compare the complete Node SPKI PEM: ImportFromPem can also accept private
        // material or ignore trailing content, neither belongs in a public field.
        if (string.Concat(((string)value["publicKeyPem"]!).Where(c => !char.IsWhiteSpace(c))) != string.Concat(pair.PublicKeyPem.Where(c => !char.IsWhiteSpace(c))))
            throw new BootstrapException("Signing identity public/private keys do not match");
        return pair;
    }
    internal static Node State(object? value)
    {
        try
        {
            var result = (Node)Contracts.Parse("BootstrapStateSchema", value)!;
            foreach (var field in new[] { "cpUrl", "cpJwksUri" })
                if (result.TryGetValue(field, out var uri)) TrustedKeys.SecureEndpoint((string)uri!, allowQuery: field == "cpJwksUri");
            if (result.GetValueOrDefault("apiKey") is string apiKey && apiKey.Any(c => c < 33 || c > 126)) throw new ArgumentException();
            if (result.GetValueOrDefault("identity") is Node identity) Identity(identity);
            return result;
        }
        catch (Exception error) when (error is ArgumentException or ValidationError or CryptographicException or BootstrapException)
        { throw new BootstrapException("Invalid bootstrap state"); }
    }
    public byte[] Encrypt(object? value)
    {
        var plain = Envelope.Utf8.GetBytes(Json.Write(State(value)));
        if (plain.Length > Envelope.Limit) throw new BootstrapException("Bootstrap state exceeds 1 MiB");
        var nonce = RandomNumberGenerator.GetBytes(12);
        var encrypted = Envelope.Crypt(true, key, nonce, plain);
        return Envelope.Utf8.GetBytes(Json.Write(Contracts.Parse("BootstrapStateEnvelopeSchema", new Node
        {
            ["v"] = 1, ["iv"] = Convert.ToBase64String(nonce),
            ["tag"] = Convert.ToBase64String(encrypted[^16..]), ["ct"] = Convert.ToBase64String(encrypted[..^16])
        })));
    }
    public Node Decrypt(ReadOnlySpan<byte> value)
    {
        try
        {
            if (value.Length > 2 * Envelope.Limit) throw new ArgumentException();
            var envelope = (Node)Contracts.Parse("BootstrapStateEnvelopeSchema", Json.Read(Envelope.Utf8.GetString(value)))!;
            var iv = Envelope.Decode((string)envelope["iv"]!); var tag = Envelope.Decode((string)envelope["tag"]!);
            var ct = Envelope.Decode((string)envelope["ct"]!);
            if (iv.Length != 12 || tag.Length != 16 || ct.Length > Envelope.Limit) throw new ArgumentException();
            return State(Json.Read(Envelope.Utf8.GetString(Envelope.Crypt(false, key, iv, [.. ct, .. tag]))));
        }
        catch (Exception error) when (error is ArgumentException or ValidationError or InvalidCipherTextException or ConfigEncryptionException or BootstrapException or System.Text.Json.JsonException)
        { throw new BootstrapException("Bootstrap state authentication failed"); }
    }
}

/// <summary>One owner per store. Operations reload state and retain no plaintext cache or background work.</summary>
public sealed class BootstrapStateStore(IStateStore store, string key)
{
    private readonly BootstrapCipher cipher = new(key);
    private readonly SemaphoreSlim gate = new(1, 1);
    private async ValueTask<Node> ReadState(CancellationToken cancellation)
    {
        var raw = await store.Load(cancellation);
        return raw is null ? new Node { ["version"] = 1 } : cipher.Decrypt(raw);
    }
    public async ValueTask<Node> Read(bool redacted = false, CancellationToken cancellation = default)
    {
        await gate.WaitAsync(cancellation);
        try
        {
            var value = await ReadState(cancellation);
            if (!redacted) return value;
            var paths = new List<string[]>();
            var result = (Node)V.Encrypt(Contracts.Get("BootstrapStateSchema"), value, (path, _) =>
            {
                paths.Add(path.Cast<string>().ToArray()); return "encrypted:redacted";
            })!;
            // This canonical object has no unions; every native sensitive path is retained.
            foreach (var path in paths)
            {
                var parent = result;
                foreach (var part in path[..^1]) parent = (Node)parent[part]!;
                parent[path[^1]] = "__redacted__";
            }
            return result;
        }
        finally { gate.Release(); }
    }
    public async ValueTask<Node> Write(IDictionary<string, object?> patch, CancellationToken cancellation = default)
    {
        var copied = new Node(patch); copied.TryAdd("version", 1); copied = BootstrapCipher.State(copied);
        await gate.WaitAsync(cancellation);
        try
        {
            var value = await ReadState(cancellation);
            foreach (var (name, field) in copied) value[name] = field;
            value = BootstrapCipher.State(value);
            await store.Save(cipher.Encrypt(value), cancellation);
            return value;
        }
        finally { gate.Release(); }
    }
    public async ValueTask Clear(CancellationToken cancellation = default)
    {
        await gate.WaitAsync(cancellation);
        try { await store.Save(cipher.Encrypt(new Node { ["version"] = 1 }), cancellation); }
        finally { gate.Release(); }
    }
    public async ValueTask<KeyPair> Identity(CancellationToken cancellation = default)
    {
        await gate.WaitAsync(cancellation);
        try
        {
            var value = await ReadState(cancellation);
            if (value.GetValueOrDefault("identity") is Node identity) return BootstrapCipher.Identity(identity);
            var pair = KeyPair.Generate();
            value["identity"] = new Node { ["privateKeyPem"] = pair.PrivateKeyPem, ["publicKeyPem"] = pair.PublicKeyPem, ["kid"] = pair.Kid };
            await store.Save(cipher.Encrypt(value), cancellation);
            return pair;
        }
        finally { gate.Release(); }
    }
}
